retrier/retrier_test.go

```diff
@@ -153,6 +153,24 @@ func TestRetrierRunFnWithInfinite(t *testing.T) {
 	}
 }
 
+func TestRetrierWithDynamicBackoff(t *testing.T) {
+	r := New([]time.Duration{0, 10 * time.Millisecond}, nil)
+	st := time.Now()
+
+	err := r.Run(genWork([]error{ErrWithBackoff(errFoo, 500*time.Millisecond)}))
+	if err != nil {
+		t.Error(err)
+	}
+	if i != 2 {
+		t.Error("run wrong number of times")
+	}
+
+	if time.Since(st) < 500*time.Millisecond {
```

Comment:
I have tried to avoid doing time-based assertions in the test suite, they tend to be quite flaky.

Reply:
Yes, the test results may be flaky, but I couldn't think of any other way to check 🤔

Reply:
Currently, I don't assert on how long it actually sleeps (I trust go's timer to work) and I simply unit-test the return value of `calcSleep` (and then the bigger tests just assert on the number of times that the method gets retried, which should be deterministic).

Perhaps for this PR you could pass `err` into `calcSleep` and move the logic into that method, so you can test that logic in a unit test of that method without any time-based assertions?